scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+import time
+from concurrent.futures import ThreadPoolExecutor
+from multiprocessing import Event, cpu_count
+from multiprocessing.synchronize import Event as EventClass
+from queue import Empty, Queue
+from typing import Callable, Generic, Hashable, TypeVar
+
+_T = TypeVar("_T", bound=Hashable)
+
+
+class _Node(Generic[_T]):
+    def __init__(self, value: _T):
+        self.value = value
+        self.counter: int = 0
+        self.dependents: list[_Node[_T]] = []
+
+
+class DAG(Generic[_T]):
+    def __init__(self, dependency_graph: dict[_T, list[_T]]):
+        """
+        Construct a directed acyclic graph from an adjacency list.
+
+        The `dependency_graph` _must not_ contain any cycles.
+        """
+
+        self._nodes: dict[_T, _Node[_T]] = {}
+        self._queue: list[_T] = []
+        self._num_finished: int = 0
+
+        for node, deps in dependency_graph.items():
+            new_node = self._get_or_insert(node)
+            new_node.counter += len(deps)
+            for dep in deps:
+                self._get_or_insert(dep).dependents.append(new_node)
+
+        self._queue.extend(node.value for node in self._nodes.values() if node.counter == 0)
+
+    def _get_or_insert(self, node: _T) -> _Node[_T]:
+        if node not in self._nodes:
+            self._nodes[node] = _Node(node)
+        return self._nodes[node]
+
+    def _finish(self, node: _T) -> None:
+        # mark the `node` as finished, which decrements the pending dependency counter on its dependents
+        # once a node reaches `0` on its counter, it is marked ready and put in the queue for processing
+        for dependent in self._nodes[node].dependents:
+            dependent.counter -= 1
+            if dependent.counter == 0:
+                self._queue.append(dependent.value)
+        self._num_finished += 1
+
+    def _is_done(self) -> bool:
+        # the number of nodes in the graph should never change
+        return self._num_finished == len(self._nodes)
+
+    def walk(self, f: Callable[[_T], None], max_tokens: int, refill_interval_s: float) -> None:
+        """
+        Process the graph in parallel.
+
+        Each node in the graph is processed only once all of its dependencies have been processed.
+
+        Concurrency is limited by the following bucket rate limiting algorithm:
+        * Processing may not begin until a token can be acquired from the bucket.
+        * There are at most `max_tokens - in_progress_tasks` in the bucket at any time.
+        * Tokens are refreshed every `refill_interval_s`.
+        """
+
+        num_cpus = cpu_count()
+        # we need one main thread + N worker threads, so `N = num_cpus - 1`,
+        # and we need at least 1 worker thread.
+        num_workers = num_cpus - 1 if num_cpus > 0 else 1
```

Comment:
if `num_cpus == 1` we get `num_workers == 0` which goes against the comment

`num_workers = max(1, num_cpus - 1)` was maybe intended?

Reply:
…but the amount of parallelism should probably be up to the caller of `walk`.

In our particular case of publishing crates, each crate is already built in parallel, so publishing 32 crates in parallel on a 32-core machine is likely a bad idea. The most wins from publishing crates in parallel will come from the non-parallelizable parts of publishing (last part of the compilation, linking, uploading, …).

I'm _guessing_ publishing 2-3 crates in parallel is the sweet-spot on the CI.

Reply:
Agreed, I capped it at 3.